Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -687,6 +687,66 @@ public struct NonBlockingFileIO: Sendable {
         }
     }
 
+    private func createDirectory0(_ path: String, mode: NIOPOSIXFileMode) throws {
+        // Fail fast if not a directory or file exists
+        do {
+            var s = stat()
+            try Posix.stat(pathname: path, outStat: &s)
+            if (S_IFMT & s.st_mode) == S_IFDIR {
+                return
+            }
+            throw IOError(errnoCode: ENOTDIR, reason: "Not a directory")
+        } catch {
+            // ignore error
+        }
+
+        // Slow path
+        var index = path.index(before: path.endIndex)
+        while index != path.startIndex && path[index] == "/" {
+            index = path.index(before: index)
+        }
+
+        while index != path.startIndex, path[index] != "/" {
+            index = path.index(before: index)
+        }
```

Comment:
Can you also explain what these two blocks are doing?

Reply:
added comments